File: Tree/236.Lowest-Common-Ancestor-of-a-Binary-Tree/236.Lowest-Common-Ancestor-of-a-Binary-Tree.ts
function lowestCommonAncestor(
  root: TreeNode | null,
  p: TreeNode | null,
  q: TreeNode | null
): TreeNode | null {
  let res = null;
  dfs(root, p, q);
  return res;

  function dfs(root: TreeNode | null, p: TreeNode | null, q: TreeNode | null) {
    if (root === null) {
      return 0;
    }
    const left = dfs(root.left, p, q);
    const right = dfs(root.right, p, q);
    const self = root === q || root === p;
    const count = left + right + Number(self);
    if (count === 2 && res === null) {
      res = root;
    }
    return count;
  }
}

File: Tree/236.Lowest-Common-Ancestor-of-a-Binary-Tree/236.Lowest-Common-Ancestor-of-a-Binary-Tree.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution:
    def lowestCommonAncestor(self, root: 'TreeNode', p: 'TreeNode', q: 'TreeNode') -> 'TreeNode':
        result = None

        def dfs(root):
            nonlocal result
            if not root:
                return 0

            left = dfs(root.left)
            right = dfs(root.right)
            
            selfVal = 1 if root.val == p.val or root.val == q.val else 0
            if result == None and left + right + selfVal == 2:
                result = TreeNode(root.val)
            return left + right + selfVal
            
        dfs(root)
        return result
        
